src/rulesets/LinearAlgebra/utils.jl

```diff
@@ -43,3 +43,36 @@ Symmetric
 ````
 """
 _unionall_wrapper(::Type{T}) where {T} = T.name.wrapper
+
+"""
+    WithSomeZeros{T}
+
+This is a union of LinearAlgebra types, all of which are partly structral zeros,
+with a simple backing array given by `parent(x)`. All have methods of `_rewrap`
+to re-create.
+
+This exists to solve a type instability, as broadcasting for instance
+`λ .* Diagonal(rand(3))` gives a dense matrix when `x==Inf`.
+But `withsomezeros_rewrap(x, λ .* parent(x))` is type-stable.
+"""
+WithSomeZeros{T} = Union{
+    Diagonal{T},
+    UpperTriangular{T},
+    UnitUpperTriangular{T},
+    # UpperHessenberg{T},  # doesn't exist in Julia 1.0
+    LowerTriangular{T},
+    UnitLowerTriangular{T},
+}
+for S in [
+    :Diagonal,
+    :UpperTriangular,
+    :UnitUpperTriangular,
+    # :UpperHessenberg,
+    :LowerTriangular,
+    :UnitLowerTriangular,
+]
+    @eval withsomezeros_rewrap(::$S, x) = $S(x)
+end
+
+# Bidiagonal, Tridiagonal have more complicated storage.
```

Comment:
For these would we still expect `parent` to work, but for `withsomezeros_rewrap(::$S, x) = $S(x)` not to work?
So we would need to work out a different definition?
Or should we introduce `unwrap` which defaults to `parent` but that has other defintions?

Reply:
One could extend a mechanism like this, but `Bidiagonal`'s underlying storage is two vectors, so it will be a bit more complicated. 

On the other hand, I've also yet to encounter a Bidiagonal in the wild, let alone one being used with AD. So perhaps we can kick the can down the road & let someone optimise it when they need it.

Reply:
Why are we defining a method for `rewrap` rather than just using the constructor,
but just use `parent` rather than defining a  `unwrap` method?

Reply:
Oh I see, yeah this is fair.
I take it sometimes the full `typeof(x)` doesn't work because the eltype might change?
(I think ideally i wouldn't)

Reply:
Fixing it to exactly the same type seems too strong, or at least nothing else does this. Can't you get a `Diagonal(Fill(1.0, 10))` in the backward pass, or something? Or like my example have integers... even if not a range. (Probably Will would say that ought to go to some Composite(Diagonal(Composite(start=... and then I have no idea what'll happen.)